File: framework/backend/python/packages/k1s0-rate-limit/src/k1s0_rate_limit/sliding_window.py
"""Sliding window rate limiter implementation."""

from __future__ import annotations

import asyncio
import time
from collections import deque
from datetime import timedelta

from k1s0_rate_limit.config import SlidingWindowConfig
from k1s0_rate_limit.rate_limiter import RateLimiter, RateLimitStats


class SlidingWindow(RateLimiter):
    """Rate limiter using the sliding window algorithm.

    Tracks individual request timestamps within a rolling window.
    Requests are rejected once ``max_requests`` is reached within
    the ``window_size`` duration.

    Args:
        config: Sliding window configuration.

    Example::

        limiter = SlidingWindow(SlidingWindowConfig(
            window_size=timedelta(seconds=60),
            max_requests=100,
        ))
        if await limiter.try_acquire():
            process_request()
    """

    def __init__(self, config: SlidingWindowConfig | None = None) -> None:
        cfg = config or SlidingWindowConfig()
        self._window_size = cfg.window_size.total_seconds()
        self._max_requests = cfg.max_requests
        self._timestamps: deque[float] = deque()
        self._lock = asyncio.Lock()
        self._allowed = 0
        self._rejected = 0

    def _cleanup(self) -> None:
        """Remove timestamps outside the current window."""
        cutoff = time.monotonic() - self._window_size
        while self._timestamps and self._timestamps[0] < cutoff:
            self._timestamps.popleft()

    async def try_acquire(self) -> bool:
        """Attempt to record a request within the current window.

        Returns:
            ``True`` if the request is within the limit, ``False`` otherwise.
        """
        async with self._lock:
            self._cleanup()
            if len(self._timestamps) < self._max_requests:
                self._timestamps.append(time.monotonic())
                self._allowed += 1
                return True
            self._rejected += 1
            return False

    def time_until_available(self) -> timedelta:
        """Return estimated wait time until a slot opens in the window.

        Returns:
            A ``timedelta`` representing the wait duration.
        """
        self._cleanup()
        if len(self._timestamps) < self._max_requests:
            return timedelta()
        if not self._timestamps:
            return timedelta()
        # Oldest timestamp will expire first
        oldest = self._timestamps[0]
        wait = (oldest + self._window_size) - time.monotonic()
        return timedelta(seconds=max(0.0, wait))

    def available_tokens(self) -> int:
        """Return the number of requests still available in the current window.

        Returns:
            Non-negative count of remaining request slots.
        """
        self._cleanup()
        return max(0, self._max_requests - len(self._timestamps))

    def stats(self) -> RateLimitStats:
        """Return a snapshot of current statistics.

        Returns:
            Rate limit statistics including allowed, rejected, and available.
        """
        self._cleanup()
        return RateLimitStats(
            allowed=self._allowed,
            rejected=self._rejected,
            total=self._allowed + self._rejected,
            available=max(0, self._max_requests - len(self._timestamps)),
        )

File: framework/backend/kotlin/packages/k1s0-rate-limit/src/main/kotlin/dev/k1s0/ratelimit/SlidingWindow.kt
package dev.k1s0.ratelimit

import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlin.time.TimeSource

private val logger = KotlinLogging.logger {}

/**
 * Sliding window rate limiter configuration.
 *
 * @property maxRequests Maximum number of requests allowed in the window.
 * @property window Duration of the sliding window.
 */
public data class SlidingWindowConfig(
    val maxRequests: Long = 1000,
    val window: Duration = 60.seconds,
)

/**
 * Sliding window rate limiter.
 *
 * Tracks request timestamps within a rolling time window and rejects
 * requests when the limit is exceeded.
 */
public class SlidingWindow(
    config: SlidingWindowConfig = SlidingWindowConfig(),
) : RateLimiter {
    private val maxRequests = config.maxRequests
    private val window = config.window
    private val timestamps = ArrayDeque<TimeSource.Monotonic.ValueTimeMark>()
    private val mutex = Mutex()
    private var allowed = 0L
    private var rejected = 0L

    private fun evictExpired() {
        val now = TimeSource.Monotonic.markNow()
        while (timestamps.isNotEmpty()) {
            val oldest = timestamps.first()
            if (oldest.elapsedNow() >= window) {
                timestamps.removeFirst()
            } else {
                break
            }
        }
    }

    override suspend fun tryAcquire(): Boolean = mutex.withLock {
        evictExpired()
        if (timestamps.size < maxRequests) {
            timestamps.addLast(TimeSource.Monotonic.markNow())
            allowed++
            logger.trace { "Request allowed (count=${timestamps.size}/$maxRequests)" }
            true
        } else {
            rejected++
            logger.debug { "Request rejected (count=${timestamps.size}/$maxRequests)" }
            false
        }
    }

    override fun timeUntilAvailable(): Duration {
        if (timestamps.isEmpty() || timestamps.size < maxRequests) {
            return Duration.ZERO
        }
        val oldest = timestamps.first()
        val elapsed = oldest.elapsedNow()
        return if (elapsed >= window) Duration.ZERO else window - elapsed
    }

    override fun availableTokens(): Long {
        evictExpired()
        return maxOf(0L, maxRequests - timestamps.size)
    }

    override fun stats(): RateLimitStats = RateLimitStats(
        allowed = allowed,
        rejected = rejected,
        total = allowed + rejected,
        available = availableTokens(),
    )
}
